### PR TITLE
Convert BME280 to multi-instance, use new DT API

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Make some sample changes as well.

Tested with two devices at the same time, one on each bus.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
```
$ wc -l /home/mbolivar/zp/zephyr/sanity-out/frdm_k64f/tests/drivers/build_all/sensors.build.a_h/build.log
78041 /home/mbolivar/zp/zephyr/sanity-out/frdm_k64f/tests/drivers/build_all/sensors.build.a_h/build.log
```

That's an impressive error message. We're going to have to do something about making these shorter.

File: drivers/sensor/bme280/bme280.c [hunk 1]
```diff
@@ -306,95 +370,187 @@ static int bme280_chip_init(struct device *dev)
...
+		    (DT_INST_SPI_DEV_CS_GPIO_LABEL(inst)), (NULL))
+
+#define BME280_CONFIG_I2C(inst)						\
+	{								\
+		.bus_label = DT_INST_BUS_LABEL(inst),			\
+		.reg_io = &bme280_reg_io_i2c,				\
+		.bus_config =  { .i2c_addr = DT_INST_REG_ADDR(inst), }	\
+	}
+
+/* TODO replace with a foreach over all enabled devices when available */
```
Comment (gmarull) [inline drivers/sensor/bme280/bme280.c]: @mbolivar-nordic when will the `foreach` macro be available?

Comment (mbolivar-nordic) [reply drivers/sensor/bme280/bme280.c]: There is a PR. When it's merged it'll be ready :).

Comment (gmarull) [reply drivers/sensor/bme280/bme280.c]: seen #23416 thanks!

**Comment by mbolivar-nordic** [comment]
Rebased on top of https://github.com/zephyrproject-rtos/zephyr/pull/23792, DNM until that has a chance to be reviewed

**Comment by mbolivar-nordic** [comment]
> Rebased on top of #23792, DNM until that has a chance to be reviewed

Rebased on master, dropping DNM

**Comment by mbolivar-nordic** [comment]
Rebased again

File: drivers/sensor/bme280/bme280.h [hunk 2]
```diff
@@ -10,6 +10,11 @@
 
 #include <zephyr/types.h>
 #include <device.h>
+#include <devicetree.h>
+
+#define DT_DRV_COMPAT bosch_bme280
```
Comment (galak) [inline drivers/sensor/bme280/bme280.h:15]: I was avoid putting DT_DRV_COMPAT in header if possible.

File: drivers/sensor/bme280/bme280.c [hunk 3]
```diff
@@ -305,113 +370,213 @@ static int bme280_read_compensation(struct bme280_data *data)
...
-DEVICE_AND_API_INIT(bme280, DT_INST_LABEL(0), bme280_init, &bme280_data,
-		    NULL, POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,
-		    &bme280_api_funcs);
+/* bme280_is_on_spi() can't be used at preprocessor time. */
+#define BME280_ON_SPI(inst) DT_INST_ON_BUS(inst, spi)
+
+/* This is only usable if BME280_ON_SPI(inst) */
+#define BME280_HAS_CS(inst) DT_INST_SPI_DEV_HAS_CS_GPIOS(inst)
+
+#define BME280_DEFINE(inst)						\
```
Comment (galak) [inline drivers/sensor/bme280/bme280.c:508]: Can you move this below `BME280_DEFINE_SPI` and `BME280_DEFINE_I2C`

**Review by galak — APPROVED** [review]
In general looks fine.  Would like to see the macro's related to BME280_DEFINE re-organized and have more comments as I imagine we'll want to point people at this as an example of how to do things.

**Comment by mbolivar-nordic** [comment]
@galak review addressed, thank you. I've also added a new dt_compat_on_bus() helper. I kept getting annoyed at having to flip between prj.conf and prj_spi.conf, so the same just default-enables the right bus now using that.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**